### PR TITLE
Added ldm (local-dependency-manager) cli tool

### PR DESCRIPTION
**Pull request by akosyakov**
Look at https://github.com/theia-ide/theia/compare/ak/ldm?expand=1#diff-cef9074793014c2436f72cb7d894def5 to learn how to use.

@JanKoehnlein, @jbicker, @spoenemann it should simplify development with local dependencies.

### CODE REVIEW

**Comment by akosyakov** [comment]
Who knows how to restart a build without pushing new commits?

**Comment by kittaakos** [comment]
One should log into Travis-CI (Github account it fine) and click on the restart build. (I have restarted this build)

**Comment by akosyakov** [comment]
@hexa00 Could you try this PR together with your https://github.com/theia-ide/theia/pull/160?

**Comment by akosyakov** [comment]
@hexa00 you don't need a separate script to watch `theia-core` with this PR, [watch:dependencies](https://github.com/theia-ide/theia/pull/156/files#diff-163ff56fd2fd53d36d314074510353c0R30) does it by passing `--run=watch` to ldm. See [README](https://github.com/theia-ide/theia/pull/156/files#diff-cef9074793014c2436f72cb7d894def5) for ldm commands and options. Also you can remove `cleanSync` by not passing `--sync` option to ldm.

**Comment by hexa00** [comment]
OK let me try that

**Comment by hexa00** [comment]
I get src/watcher.ts(8,22): error TS2307: Cannot find module 'cpx'.

**Comment by hexa00** [comment]
in cd browser && npm install
I'll try more after lunch

**Comment by hexa00** [comment]
Installing cpx in theia-core fixed that issue... testing now..

**Comment by akosyakov** [comment]
If you run ldm directly from cli then cpx should be installed globally not in theia-core. If you run watch:dependencies on example, you should run npm install first there to install ldm in local node_modules.

**Comment by hexa00** [comment]
Yes I did npm install in the browser example and then tried npm run watch:dependencies but that did not work... cpx is not a dependency that should be fixed ?  as a dev dependency in theia core ?

**Comment by hexa00** [comment]
ls -l  ./node_modules/local-dependency-manager/bin/ldm
-rw-r--r-- 1 eantotr eusers 44 May 26 12:42 ./node_modules/local-dependency-manager/bin/ldm
ldm is not executable... 
Thus running npm run watch:dependencies yields:
```> ldm watch --sync --verbose --run=watch

sh: 1: ldm: Permission denied```

**Comment by hexa00** [comment]
I tried importing my logging commit over and then running  npm run watch:dependencies (with --sync) 
But it's not copying the new files, I get:

```
 tsc --watch

src/backend/main.ts(20,37): error TS2307: Cannot find module 'theia-core/lib/application/node/logger-backend-module'.
src/backend/main.ts(21,36): error TS2307: Cannot find module 'theia-core/lib/application/node/logger-server-module'.
src/frontend/main.ts(18,38): error TS2307: Cannot find module 'theia-core/lib/application/browser/logger-frontend-module'.
2:11:30 PM - Compilation complete. Watching for file changes
```

The files are in theia-core/lib:

```
theia/lib/application/node git:(test-ldm)  ls *logger*
logger-backend-module.d.ts  logger-backend-module.js.map  logger-server-module.js
logger-backend-module.js    logger-server-module.d.ts     logger-server-module.js.map
```

But not in examples/browser/node_modules:

```
theia/examples/browser/node_modules/theia-core/lib/application/node git:(test-ldm)  ls *logger*
zsh: no matches found: *logger*
```

**Comment by akosyakov** [comment]
We should try [yalc](https://www.npmjs.com/package/yalc) instead of implementing our own tool.

**Comment by akosyakov** [comment]
with npm5 local path dependencies will use symlinks on npm install: https://github.com/npm/npm/pull/15900#issuecomment-304437353 🙈

**Comment by akosyakov** [comment]
1. I've played with yalc, but it does not allow to update dependencies from the dependent project and do it continuously.
2. I've thought about requirements to such tool:
  - it should not be necessary to update package.json to switch between local and remote dependencies
  - the same node_modules structure should be used for local and remote dependencies
  - it should be possible to update local dependencies from the dependent package and synchronize changes if it is required
    - it should be supported in instant and continuous (watch) modes
    - it should be incremental
    - it should respect npm packing: https://docs.npmjs.com/files/package.json#files
3. ldm addresses them:
  - it does not rely on local path and file versions, but on `localDependencies` and `localDevDependencies`properties in `package.json` and additional scripts like `localinstall` for development with local dependencies
  - it uses `npm pack` and `npm install` from the archived package to install dependencies instead of `npm link` or `npm install` from local path dependencies which involve symlinks (local paths since npm5)
  - it allows running any scripts on local dependencies in instant with `ldm run build` and continuous mode with `ldm watch --run=watch`
4. things to improve:
  - [ ] add a  checksum check before `cleanSync` to avoid unnecessary copying
  - [ ] reflect npm packaging logic to find files to watch
  - [ ] consider to reuse vscode logic to watch files for ldm as well for theia, I see that even though they use chokidar for unix they do additional stuff to buffer and preprocess events, recognize renames and so on, they also use another watcher implementation for windows. Luckily, it is a nicely isolated piece of code that actually can be ran in a separated process.
5. @hexa00 I've found some bugs related to dir watching, could you try again?

**Comment by hexa00** [comment]
OK I did more tests this morning and I'm happy to report that it works much better now !
Here's what I did:

```npm install  && npm install -g``` in ldm directory

Then in the browser example 
```
npm run watch:localdeps
npm run watch:backend
npm run watch:frontend
``` 
Now try to change files in backend code and frontend code, changes are reflected properly, even importing my logging patch (after an npm install of the deps)

I tried with terminial-backend/frontend/widget all worked fine.

I did stumble upon a problem however, running ```npm run watch:frontend``` it seems the webpack watcher can see imcomplete files as they are copied by ldm and we get back into the

 ```
SyntaxError: Unexpected end of JSON input
    at JSON.parse (<anonymous>)
    at /home/eantotr/ownCloud/work/theia/examples/browser/node_modules/source-map-loader/index.js:44:22
    at tryToString (fs.js:449:3)
    at FSReqWrap.readFileAfterClose [as oncomplete] (fs.js:436:12)
````

Type of error, but it can easily be fixed with a restart of the watch and it doesn't happen all the time, I did a test just stopping and starting ```npm run watch:localdeps``` with a ```npm run watch:frontend``` running and it happens around 20% of the time... 

And I could not trigger it just manually editing a file.

So thanks @akosyakov  !! this makes the work flow possible :)

BTW I agree about your requirements and that yalc was not the way to go for that

**Comment by hexa00** [comment]
Humm I got this error again:
```
undefined:1

SyntaxError: Unexpected end of JSON input
    at JSON.parse (<anonymous>)
    at /home/eantotr/ownCloud/work/theia/examples/browser/node_modules/source-map-loader/index.js:44:22
    at tryToString (fs.js:449:3)
    at FSReqWrap.readFileAfterClose [as oncomplete] (fs.js:436:12)
```

But this time I could not recover by just restarting npm run watch:frontend or restarting npm run watch:localdeps :(

**Comment by akosyakov** [comment]
@hexa00 yes, there are issues with chokidar and, consequently, with [cpx](https://github.com/mysticatea/cpx/issues/29#issuecomment-304775161) and our filesystem, I would like to tackle them separately, probably with reusing vscode code.

I rarely use `watch` from `examples`, more `build`, and it is important for me that on `build` is up to date code from dependencies is used. With this PR, I use `watch` only on `theia-core`, since `ldm` syncs changes from local dependencies on a build of the browser example.

It could be different though if we try to break `theia-core` to individual extension packages with `ldm`. I also would like to do it separately from this PR.

Could somebody else review it to decide how we proceed with it?

**Comment by hexa00** [comment]
@akosyakov Agreed we need to tackle that separately

Actually do you think it would be a good idea to just run npm run build:backend && npm run build::frontend after an ldm run ? e.g by allowing ldm to run an npm command after it's done a sync

This would avoid the watching problem by introducing a clear sequence ?

**Comment by akosyakov** [comment]
> Actually do you think it would be a good idea to just run npm run build:backend && npm run build::frontend after an ldm run ? e.g by allowing ldm to run an npm command after it's done a sync

it is already done by `build` script in the browser example: https://github.com/theia-ide/theia/pull/156/files#diff-163ff56fd2fd53d36d314074510353c0R25.

File: config/local-dependency-manager/src/package.ts [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+                }
+            }
+        }
+    }
+
+    uninstallDependency(dependency: Package): void {
+        this.execSync('npm', 'uninstall', dependency.name);
+    }
+
+    cleanDependency(dependency: Package): void {
```
Comment (akosyakov) [inline config/local-dependency-manager/src/package.ts]: uninstall remove all transient dependencies as well, it is more expensive

Comment (svenefftinge) [inline config/local-dependency-manager/src/package.ts]: ah, ok

File: config/local-dependency-manager/src/package.ts [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+            return this.raw.files;
+        }
+        return ['lib', 'src'];
+    }
+
+    resolvePath(localPath: string): string {
+        return path.normalize(path.resolve(this.packagePath, localPath));
+    }
+
+    getNodeModulePath(dependency: undefined): undefined;
```
Comment (akosyakov) [inline config/local-dependency-manager/src/package.ts]: I will remove there are no real clients for it, but it mostly to ensure that if you call with defined value then you never get `undefined`.

**Pull request merged**